cpp/cograph_recognition/cograph/Recognition.cpp

```diff
@@ -0,0 +1,233 @@
+        this.outEdges = outEdges;
+    }
+    vector<CoNode> getoutEdges(){
+        return outEdges;
+    }
+    CoNode* getParent(){
+        return parent;
+    }
+    CoNode* setParent(CoNode *parent){
+    this.parent = parent;
+    }
+    void set_md(int md){
+        this.md = md;
+    }
+    void inc_md(){
+        this.md++;
+    }
+    int get_md(){
+        return md;
+    }
+    void set_d(int d){
+            this.d = d;
+        }
+        void inc_d(){
+            this.d++;
+        }
+        int get_d(){
+            return d;
+        }
+    Marked Marked_or_not(){
+        return marked;
+    }
+    void unmark_for_new_iteration(){
+    marked = Marked::UNMARKED;
+    md = 0;
+    }
+    void mark(){
+    marked = Marked::MARKED;}
+    void unmark(){
+        marked = Marked::MARKED_AND_UNMARKED;
+    }
+}
+class CoTree{
+    private:
+    CoNode* root;
+    public:
+    CoTree(CoNode* root):root(root){
+    }
+    CoNode* getRoot(){
+        return root;
+    }
+    void setRoot(CoNode* R){
+        root = R;
+    }
+
+};
+CoTree T;
+NetworKit::Graph G;
+int mark_count = 0;
+int mark_and_unmarked_count = 0;
+int mark_ever_count = 0;
+void unmark(queue<CoNode> &marked_with_d_equal_to_md){
+    CoNode u = marked_with_d_equal_to_md.front();
+    marked_with_d_equal_to_md.pop();
+    u.unmark();
+    mark_count--;
+    mark_and_unmarked_count++;
+    u.set_md(0);
+    if(u != T.getRoot()){
+        w = u.getParent();
+        w.inc_md();
+        if(w.get_md() == w.get_d()){
+            marked_with_d_equal_to_md.push(w);
+            auto l = u -> next;
+            if(u.prev != nullptr){
+                (u.prev) -> next = l;
+                u.prev = nullptr;
+                u.next = w.head_of_list_of_children;
+                w.head_of_list_of_children.prev = u;
+                w.head_of_list_of_children = u;
+            }
+            //else u is head
+         }
+    }
+}
+void Mark( CoNode &x){
+    queue<CoNode> marked_with_d_equal_to_md;
+    mark_count = 0;
+    mark_and_unmarked_count = 0;
+    mark_ever_count = 0;
+    for(auto u : T.outEdges(x)){
+        u.mark();
+        mark_ever_count++;
+        mark_count++;
+        marked_with_d_equal_to_md.push(u);
+    }
+    while(!marked_with_d_equal_to_md.empty()){
+        unmark( marked_with_d_equal_to_md);
+    }
+        if(mark_count){
+            if(T.getRoot().d == 1){
+                T.getRoot().mark();
+            }
+        }
+
+}
+
+void Reset_All_CoNodes(CoNode *x){
+    x -> unmark_for_new_iteration();
+    auto y = x -> head_of_list_of_children;
+    while(y != nullptr){
+        Reset_All_CoNodes(y);
+        y = y -> next;
+    }
+}
+CoNode Find_Lowest( ){
+
+}
+void Insert_x_to_CoTree(CoNode *u, CoNode *x){
+    if(u -> )
+}
+void Cograph_Recognition(NetworKit::Graph &graph){
+    G = graph;
+    vector<node>vertex;
+    vector<CoNode>covertex;
+    vector<int>pos(G.z + 1);
+    for(int i = 0; i <= G.z; i++){
```

Comment:
Please use foreach loop if you have iterate over vector or any other STL structure